---
layout: post
title: "未使用模块的Tensorflow转换问题"
date: 2025-03-11 15:02:35 +0800
description: "自定义模型存在冗余的情况有时候再tensorflow中不会被建立build，推理的时候也存在问题；方法中未被调用，导致它们的变量未被追踪。模型中定义了冗余的层或分支，但这些层在。解决方式是屏蔽相关的冗余神经网络结构。"
keywords: "未使用模块的Tensorflow转换问题"
categories: ['深度学习部署模型转移']
tags: ['人工智能', 'Tensorflow', 'Python']
artid: "146179241"
image:
    path: https://api.vvhan.com/api/bing?rand=sj&artid=146179241
    alt: "未使用模块的Tensorflow转换问题"
render_with_liquid: false
featuredImage: https://bing.ee123.net/img/rand?artid=146179241
featuredImagePreview: https://bing.ee123.net/img/rand?artid=146179241
cover: https://bing.ee123.net/img/rand?artid=146179241
image: https://bing.ee123.net/img/rand?artid=146179241
img: https://bing.ee123.net/img/rand?artid=146179241
---

<div class="blog-content-box">
 <div class="article-header-box">
  <div class="article-header">
   <div class="article-title-box">
    <h1 class="title-article" id="articleContentId">
     未使用模块的Tensorflow转换问题
    </h1>
   </div>
  </div>
 </div>
 <article class="baidu_pl">
  <div class="article_content clearfix" id="article_content">
   <link href="../../assets/css/kdoc_html_views-1a98987dfd.css" rel="stylesheet"/>
   <link href="../../assets/css/ck_htmledit_views-704d5b9767.css" rel="stylesheet"/>
   <div class="htmledit_views" id="content_views">
    <p>
     自定义模型存在冗余的情况有时候再tensorflow中不会被建立build，推理的时候也存在问题；
    </p>
    <p>
     模型中定义了冗余的层或分支，但这些层在
     <code>
      call()
     </code>
     方法中未被调用，导致它们的变量未被追踪。
    </p>
    <pre><code class="language-bash">Traceback (most recent call last):
  File "&lt;path_to_your_project&gt;/models/tf.py", line 1078, in &lt;module&gt;
    main(opt)
  File "&lt;path_to_your_project&gt;/models/tf.py", line 1073, in main
    run(**vars(opt))
  File "&lt;path_to_your_project&gt;/models/tf.py", line 1049, in run
    keras_model.summary()
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/training.py", line 3234, in summary
    layer_utils.print_summary(
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/utils/layer_utils.py", line 430, in print_summary
    print_layer(layer)
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/utils/layer_utils.py", line 400, in print_layer
    print_layer_summary_with_connections(layer, nested_level)
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/utils/layer_utils.py", line 387, in print_layer_summary_with_connections
    layer.count_params(),
    ^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 2168, in count_params
    return layer_utils.count_params(self.weights)
                                    ^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 1351, in weights
    return self.trainable_weights + self.non_trainable_weights
           ^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 1308, in trainable_weights
    children_weights = self._gather_children_attribute(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 3294, in _gather_children_attribute
    return list(
           ^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 3296, in &lt;genexpr&gt;
    getattr(layer, attribute) for layer in nested_layers
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 2276, in trainable_variables
    return self.trainable_weights
           ^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/training.py", line 2733, in trainable_weights
    trainable_variables += trackable_obj.trainable_variables
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 2276, in trainable_variables
    return self.trainable_weights
           ^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 1308, in trainable_weights
    children_weights = self._gather_children_attribute(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 3294, in _gather_children_attribute
    return list(
           ^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 3296, in &lt;genexpr&gt;
    getattr(layer, attribute) for layer in nested_layers
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/base_layer.py", line 2276, in trainable_variables
    return self.trainable_weights
           ^^^^^^^^^^^^^^^^^^^^^^
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/training.py", line 2728, in trainable_weights
    self._assert_weights_created()
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/sequential.py", line 517, in _assert_weights_created
    super(functional.Functional, self)._assert_weights_created()
  File "&lt;conda_environment_path&gt;/Lib/site-packages/keras/engine/training.py", line 3540, in _assert_weights_created
    raise ValueError(
ValueError: Weights for model 'sequential_3' have not yet been created. Weights are created when the model is first called on inputs or `build()` is called with an `input_shape`.
</code></pre>
    <p>
     解决方式是屏蔽相关的冗余神经网络结构
    </p>
    <p>
    </p>
   </div>
  </div>
 </article>
 <p alt="68747470733a2f2f:626c6f672e6373646e2e6e65742f753031333539303332372f:61727469636c652f64657461696c732f313436313739323431" class_="artid" style="display:none">
 </p>
</div>


